### ISSUE TITLE
Support for HorizontalPodAutoscaler through K8s external metrics framework

### ISSUE THREAD

**Issue by sandys** [dask/dask-kubernetes#168]
Kubernetes has had support for HorizontalPodAutoscaler that Dask can hook into and avoid building its own autoscaling code. This is done using the metrics server api - https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-metrics-apis

Here's a quick example of doing this by using GPU metrics to autoscale - https://www.alibabacloud.com/blog/auto-scaling-kubernetes-clusters-based-on-gpu-metrics_594314

You can also use external services entirely to tell K8s how to scale e.g. https://www.datadoghq.com/blog/autoscale-kubernetes-datadog/#create-the-horizontal-pod-autoscaler

This system is also supported by all the hosted K8s services - e.g. https://github.com/Azure/azure-k8s-metrics-adapter and https://aws.amazon.com/blogs/opensource/horizontal-pod-autoscaling-eks/

The good thing is that Autoscaling based on Prometheus compatible metrics is already widely supported e.g. https://www.ibm.com/support/knowledgecenter/en/SSBS6K_3.1.1/manage_cluster/hpa.html

Since Dask has prometheus endpoints, can you not build support for HPA using custom prometheus metrics ? (or not use prometheus and your own metric server).

**Comment by jacobtomlinson** [comment]
Thanks for raising this. It's a question that comes up a lot.

> Since Dask has prometheus endpoints, can you not build support for HPA using custom prometheus metrics ?

We could, but don't feel like we should.

Dask distributed is designed to be deployed on a variety of different schedulers including batch schedulers and ad-hoc clusters. Therefore it has mature scaling logic for handling adding and removing workers on all platforms. We have to build things in line with the lowest common denominator, many batch systems do not support scaling logic like HPA and so Dask has to handle that itself.

This leads us to a decision when implementing something like `dask-kubernetes`. Do we use the scaling logic within Dask, or the scaling logic within Kubernetes. We've made the decision that using the same scaling logic across all platforms helps us reuse code and be consistent when it comes to debugging. Instead of having to support multiple sets of scaling logic which work differently on each platform we support just one.

Given all of this, are there things that you feel the HPA could do that Dask is not currently doing?

**Comment by sandys** [comment]
hi @jacobtomlinson thanks for replying. 

This is really a turtles-all-the-way-down discussion and I'll try to stay as respectfully away from thin ice as i can.

But here's the thing about HPA - it does not carry scheduling logic (beyond some defaults). You plug into it and drive your own scheduling logic. We are simply talking about which API to expose here. 
For example in case you wanted to implement a totally new scheduling algorithm, you could do it using the Kubernetes Scheduler Extender framework - https://developer.ibm.com/articles/creating-a-custom-kube-scheduler/

For example, if you detect you are running on batch cluster - you expose a certain metric that is consumed by HPA to schedule. If you detect an ad-hoc cluster, then you expose a different metric that is consumed by HPA. You achieve the same end-goal, but you do that by integrating with the larger k8s ecosystem (which gives you nice things like HPA loggers, dashboards on the cloud, etc etc). You do the same with Scheduler Extenders.

You guys control the scope of what Dask ought to be - I think it is overkill to run on k8s and not leverage its battle-tested primitives. But the k8s platform has specifically been built to be extended.

Tl;dr - Dask should definitely have its own scheduler logic and algorithms. Use kubernetes extendability apis to do it rather than entirely separately. Helps end users leverage the existing k8s ecosystem

**Comment by jacobtomlinson** [comment]
You make some interesting points that I would be keen to explore further.

Dask workers are semi-stateful. By that I mean they hold state in memory and if a worker is lost some amount of work is required to rebuild the state. Depending on the code you are running this could be a substantial amount of work. Therefore it is important for the Dask scheduler to be able to manage the memory of workers and drain them before removing them.

This was the reason why we moved away from `Deployment` resources. The dask scheduler would drain a worker, the worker process would exit cleanly and the scheduler would reduce the replica count. This resulted in a race condition where the worker may restart automatically and then kubernetes would remove a different worker which would cause state to be lost. This is probably avoidable with advanced configuration, but we wanted to absorb that complexity in Dask.

> I think it is overkill to run on k8s and not leverage its battle-tested primitives

As I said we are deliberately sharing as much logic as possible and are limited by the lowest common denominator. Why have ten code pathways for this when we can have one.

> which gives you nice things like HPA loggers, dashboards on the cloud, etc etc

I can appreciate this and would be happy to explore how we can integrate better with these things. Provided we do not step in the principles above.

> Helps end users leverage the existing k8s ecosystem

Kubernetes ecosystems are often diverse and have a broad range of maturity. I completely agree that  we should strive to support advanced users functionality, but that can't be at the expense of increasing beginner complexity or too much maintenance overhead.
